### ISSUE TITLE
High CPU load when running in staging environment

### ISSUE THREAD

**Issue by tompesman** [ileitch/rapns#28]
Hello,

I have 2 rapns processes running on the same server, one process runs in the production environment, the other one runs in the staging environment. In the production environment the CPU load is 4-5% and in the staging environment the load is around 25%. I suspect the logging of the SQL requests in the staging.log file increases the CPU load. Is this the expected behavior? Can I control the level of logging from the configuration file? If not, you can consider this issue a feature request. 

Regards,

Tom

**Comment by ileitch** [comment]
Even 4-5% sounds very high. Can you please paste your rapns.yml?

**Comment by tompesman** [comment]
basically the default configuration file with the production part copied for the staging environment:

``` yml
development:
  certificate: development.pem
  certificate_password:

  push:
    host: gateway.sandbox.push.apple.com
    port: 2195
    poll: 2
    connections: 3

  feedback:
    host: feedback.sandbox.push.apple.com
    port: 2196
    poll: 60

staging:
  certificate: staging.pem
  certificate_password:
  airbrake_notify: false
  pid_file: tmp/pids/rapns.pid

  push:
    host: gateway.push.apple.com
    port: 2195
    poll: 2
    connections: 3

  feedback:
    host: feedback.push.apple.com
    port: 2196
    poll: 60

production:
  certificate: production.pem
  certificate_password:
  airbrake_notify: false
  pid_file: tmp/pids/rapns.pid

  push:
    host: gateway.push.apple.com
    port: 2195
    poll: 2
    connections: 3

  feedback:
    host: feedback.push.apple.com
    port: 2196
    poll: 60
```

**Comment by ileitch** [comment]
Are you using v1.0.2? What Ruby and Rails versions? How many notifications are you sending a minute?

**Comment by tompesman** [comment]
Yes, using version: 1.0.2
ruby -v: ruby 1.8.7 (2011-02-18 patchlevel 334) [i686-linux], MBARI 0x8770, Ruby Enterprise Edition 2011.03
Ruby on Rails version 3.1.3

In the staging environment the notifications per minute vary between 0 and 10. The environment/staging.rb is copied from the production.rb file.

**Comment by ileitch** [comment]
How many notifications are you sending in production? Does the CPU load sit stable at 25% regardless of the number of notifications?

Logging is a possibility, there are improvements that could be made. Threading may be another issue since you are running 1.8, but still 25% seems higher than I would attribute to either logging or threading.

I will need to do some perf testing to see if I can identify the bottle neck.

**Comment by tompesman** [comment]
I've disabled logging with the following command in the bin/rapns file, on line 27: 

``` ruby
ActiveRecord::Base.logger.level = Logger::Severity::UNKNOWN
```

The load goes down to 20%, note that 20% is the minimum, when sending messages, the load goes up. Maybe the logging is not the problem.

Staging: 0 messages load -> 20%
Production: 0 messages load -> 5%

**Comment by tompesman** [comment]
Found the source of the problem. Every 2 seconds the following 4 queries are issued to the database.

```
  Rapns::Notification Load (302.5ms)  SELECT `rapns_notifications`.* FROM `rapns_notifications` WHERE (deliver_after IS NULL)
  Rapns::Notification Load (33.4ms)  SELECT `rapns_notifications`.* FROM `rapns_notifications` WHERE (deliver_after < '2012-01-09 11:57:18')
  Rapns::Notification Load (2.4ms)  SELECT `rapns_notifications`.* FROM `rapns_notifications` WHERE `rapns_notifications`.`failed` = 0 AND `rapns_notifications`.`delivered` = 0
  Rapns::Notification Load (206.9ms)  SELECT `rapns_notifications`.* FROM `rapns_notifications` 
```

The notifications table contains over 2500 records, so the last query might be the problem? I've cleared the notifications table and the load is down again ( less 1% CPU).

**Comment by ileitch** [comment]
Ah thanks for looking into this. I'll take a look at the indexes as soon as I can, very busy at the moment.

**Comment by ileitch** [comment]
Should be fixed in 1.0.3.
